### PR TITLE
HIVE-27118: Add HMS token to UGI,for iceberg table insert on HiveClient

### PR DESCRIPTION
**Pull request by hzxiongyinke**
### What changes were proposed in this pull request?
Fix iceberg table insert on hive client throw GSS initiate failed exception, for issue https://github.com/apache/iceberg/issues/3127

### Why are the changes needed?
add HiveMetastore Token to UGI, so HiveMetastoreClient on yarn do not need kerberos init angine.

### Does this PR introduce _any_ user-facing change?
No

### Is the change a dependency upgrade?
No

### How was this patch tested?
UT

### CODE REVIEW

**Comment by hzxiongyinke** [comment]
@pvary Can you help me take a look? thanks

**Comment by pvary** [comment]
@hzxiongyinke: I am not actively working on Hive lately. You might want to find some other reviewers.

Also, it is a good practice to provide a unit test for your fixes which fails before the patch, and works after the patch. This would prevent future developers braking your fix.

**Comment by hzxiongyinke** [comment]
> @hzxiongyinke: I am not actively working on Hive lately. You might want to find some other reviewers.
> 
> Also, it is a good practice to provide a unit test for your fixes which fails before the patch, and works after the patch. This would prevent future developers braking your fix.

@pvary This Issue 

> @hzxiongyinke: I am not actively working on Hive lately. You might want to find some other reviewers.
> 
> Also, it is a good practice to provide a unit test for your fixes which fails before the patch, and works after the patch. This would prevent future developers braking your fix.

@pvary This  PR is related to Hive client writing to Iceberg table when HiveMetastore is enabled with Kerberos authentication. During the commit phase of inserting into the Iceberg table using Hive client, an error occurs. The related issue is https://github.com/apache/iceberg/issues/3127, but I don't know how to add this test case.

**Comment by hzxiongyinke** [comment]
@SourabhBadhya Can you help me take a look? How should I add this qtest?thks！

**Comment by SourabhBadhya** [comment]
Hi @hzxiongyinke,
Consider adding tests in the files where SessionState is tested. A potential candidate for adding such tests is `TestSessionState`. But feel free to add tests at relevant places as you wish.

I dont think QTests is going to help much here, since it's primarily used for query output correctness. But let me know if you need help on writing QTests for this patch.

**Comment by hzxiongyinke** [comment]
> SourabhBadhya
Thank you for your response, but there still have a few questions:

- If I add test cases in `TestSessionState`, do I need to start a remote HiveMetastore and enable related configurations such as Kerberos?

- What should be the granularity of my testing?

**Comment by SourabhBadhya** [comment]
@hzxiongyinke , 
I would say it would be great to have integration test but having unit tests around whether the HMS token is added to the UGI would also be fine.

You can simulate a lot of these logic by first constructing `conf` objects and adding the desired properties to it as needed and then expecting that in the UGI. You can also mock objects and make it pass through the logic and assert the expected method calls / values.

**Comment by hzxiongyinke** [comment]
> @hzxiongyinke , I would say it would be great to have integration test but having unit tests around whether the HMS token is added to the UGI would also be fine.
> 
> You can simulate a lot of these logic by first constructing `conf` objects and adding the desired properties to it as needed and then expecting that in the UGI. You can also mock objects and make it pass through the logic and assert the expected method calls / values.

@SourabhBadhya 
I have finished writing my unit test cases, which are located in the directory itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/session. I plan to use the command "mvn test -Dtest=xxx" to test them, but I found that the Test Cases were not triggered. Later, I discovered that the maven-surefire-plugin was missing in the hive-unit pom.xml. Is what I am doing correct?

**Comment by SourabhBadhya** [comment]
You must run the test inside `itests/hive-unit` directory using the command - 
`mvn test -Dtest=<TestName>`

Again all of this information is available here - https://cwiki.apache.org/confluence/display/hive/hivedeveloperfaq as stated earlier. Refer this.

**Comment by hzxiongyinke** [comment]
> You must run the test inside `itests/hive-unit` directory using the command - `mvn test -Dtest=<TestName>`
> 
> Again all of this information is available here - https://cwiki.apache.org/confluence/display/hive/hivedeveloperfaq as stated earlier. Refer this.

@SourabhBadhya 
I have identified the cause of the problem. The junit dependency in the pom.xml file of hive-unit is incorrect. After applying patch HIVE-23482, the maven.surefire.plugin.version was upgraded to 3.0.0-M4 to enable junit5.

So when I execute "mvn test -Dtest=<TestName>" under itests/hive-unit, no test cases are triggered because the junit pom dependency is incorrect.

File: ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java [hunk 1]
```diff
@@ -731,6 +734,24 @@ private static void start(SessionState startSs, boolean isAsync, LogHelper conso
       // that would cause ClassNoFoundException otherwise
       throw new RuntimeException(e);
     }
+    try {
+      if(!startSs.isHiveServerQuery){
```
Comment (SourabhBadhya) [inline ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java:738]: A lot of nested if conditionals. Can we move this inside a method and probably invert the if conditions so that if those are satisfied then return. 
Something like - 
```
void methodName() {
       if (startSs.isHiveServerQuery) {
           return;
       }
       if (!getSessionConf().getBoolVar(ConfVars.METASTORE_USE_THRIFT_SASL) {
           return;
       }
       ...... (so on)
}
```

File: ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java [hunk 2]
```diff
@@ -731,6 +734,24 @@ private static void start(SessionState startSs, boolean isAsync, LogHelper conso
...
+            getSessionConf().setVar(ConfVars.METASTORE_TOKEN_SIGNATURE, CLITOKEN);
+            try {
+              SecurityUtils.setTokenStr(UGI, hmsDelegationTokenStr, CLITOKEN);
+            } catch (IOException e) {
+              throw new IOException("Couldn't setup delegation token in the ugi: " + e, e);
+            }
+          }
+        }
+      }
+    }catch (Exception e) {
```
Comment (SourabhBadhya) [inline ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java:752]: Nit: Space before catch.

File: ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java [hunk 3]
```diff
@@ -731,6 +734,24 @@ private static void start(SessionState startSs, boolean isAsync, LogHelper conso
...
+      if(!startSs.isHiveServerQuery){
+        if (getSessionConf().getBoolVar(ConfVars.METASTORE_USE_THRIFT_SASL)) {
+          UserGroupInformation UGI = Utils.getUGI();
+          String hmsDelegationTokenStr = Hive.get().getDelegationToken(UGI.getShortUserName(), UGI.getShortUserName());
+          if (hmsDelegationTokenStr != null) {
+            getSessionConf().setVar(ConfVars.METASTORE_TOKEN_SIGNATURE, CLITOKEN);
+            try {
+              SecurityUtils.setTokenStr(UGI, hmsDelegationTokenStr, CLITOKEN);
+            } catch (IOException e) {
+              throw new IOException("Couldn't setup delegation token in the ugi: " + e, e);
```
Comment (SourabhBadhya) [inline ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java:747]: I think this is better - 
`throw new IOException("Couldn't set delegation token in the UGI due to: " + e.getMessage(), e);`

File: itests/hive-minikdc/src/test/java/org/apache/hadoop/hive/ql/session/TestAddMetastoreDelegationTokenToUGI.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+      String hiveMetastoreKeytab = miniKDC.getKeyTabFile(
+              miniKDC.getServicePrincipalForUser(miniKDC.getHiveMetastoreServicePrincipal()));
+
+      initConf();
+      MetastoreConf.setBoolVar(conf, ConfVars.USE_THRIFT_SASL, true);
+      MetastoreConf.setVar(conf, ConfVars.KERBEROS_PRINCIPAL, hiveMetastorePrincipal);
+      MetastoreConf.setVar(conf, ConfVars.KERBEROS_KEYTAB_FILE, hiveMetastoreKeytab);
+      MetastoreConf.setBoolVar(conf, ConfVars.EXECUTE_SET_UGI, false);
+    }
+    super.setUp();
```
Comment (SourabhBadhya) [inline itests/hive-minikdc/src/test/java/org/apache/hadoop/hive/ql/session/TestAddMetastoreDelegationTokenToUGI.java:63]: Isn't this same code as what is present in `setup()` method of `TestRemoteHiveMetaStoreKerberos.java`? Consider reusing the setup method. Or better move the test there itself.

**Review by SourabhBadhya — CHANGES_REQUESTED** [review]

File: itests/hive-minikdc/src/test/java/org/apache/hadoop/hive/ql/session/TestAddMetastoreDelegationTokenToUGI.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+import org.junit.Test;
+import static org.junit.Assert.assertNotNull;
+
+
+import java.util.List;
+
+/**
+ * TestAddMetastoreDelegationTokenToUGI : Simple base test to add metastore delegation token to UserGroupInformation object
+ */
+public class TestAddMetastoreDelegationTokenToUGI extends TestRemoteHiveMetaStore {
```
[the remark this answers is not in the document]
Comment (hzxiongyinke) [reply itests/hive-minikdc/src/test/java/org/apache/hadoop/hive/ql/session/TestAddMetastoreDelegationTokenToUGI.java:43]: > Can we move this test to itests/hive-minikdc? You seem to use that package a lot in this test and getting the utils will be easier.

What you mean is that TestAddMetastoreDelegationTokenToUGI is located under itest/hive-minikdc/src/test/java/org/apache/hive/minikdc, instead of itest/hive-minikdc/src/test/java/org/apache/hadoop/hive/ql/session? Or should we add the mentioned test case directly to itest/hive-minikdc/src/test/java/org/apache/hive/minikdc/TestRemoteHiveMetaStoreKerberos.java?

File: itests/hive-minikdc/src/test/java/org/apache/hadoop/hive/ql/session/TestAddMetastoreDelegationTokenToUGI.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    super.setUp();
+  }
+
+  @Test
+  public void testGetDelegationTokenFromUGI() throws Exception {
+    HiveConf clientConf = new HiveConf(conf, TestAddMetastoreDelegationTokenToUGI.class);
+    HiveConf.setVar(clientConf, HiveConf.ConfVars.METASTOREURIS, "thrift://localhost:" + port);
+    HiveConf.setBoolVar(clientConf, HiveConf.ConfVars.METASTORE_USE_THRIFT_SASL, true);
+    SessionState.start(new CliSessionState(clientConf));
+    assertNotNull(SecurityUtils.getTokenStrForm(CLITOKEN));
```
[the remark this answers is not in the document]
Comment (hzxiongyinke) [reply itests/hive-minikdc/src/test/java/org/apache/hadoop/hive/ql/session/TestAddMetastoreDelegationTokenToUGI.java:72]: > Can we add another test where without the following setup, the CLITOKEN retrieved is null.

I will be adding another test case.

File: ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java [hunk 7]
```diff
@@ -131,6 +132,8 @@ public class SessionState implements ISessionAuthState{
   private static final String LOCAL_SESSION_PATH_KEY = "_hive.local.session.path";
   private static final String HDFS_SESSION_PATH_KEY = "_hive.hdfs.session.path";
   private static final String TMP_TABLE_SPACE_KEY = "_hive.tmp_table_space";
+  private static final String CLITOKEN = "HiveClientImpersonationToken";
```
[the remark this answers is not in the document]
Comment (hzxiongyinke) [reply ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java:135]: > Is there a reason why this constant has the given name? Or this is just a custom name?

The usage of CLITOKEN here is the same as that of HS2TOKEN below.
https://github.com/apache/hive/blame/d2d4b86433ef356bf738edd72bfef74511a15444/service/src/java/org/apache/hive/service/cli/session/HiveSessionImplwithUGI.java#L45

File: ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java (hunk 2 above)
Comment (hzxiongyinke) [reply ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java:752]: > Nit: Space before catch.

fixed

File: ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java (hunk 3 above)
Comment (hzxiongyinke) [reply ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java:747]: > I think this is better - `throw new IOException("Couldn't set delegation token in the UGI due to: " + e.getMessage(), e);`

You're right,I will follow your suggestion.

File: ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java (hunk 1 above)
Comment (hzxiongyinke) [reply ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java:738]: > A lot of nested if conditionals. Can we move this inside a method and probably invert the if conditions so that if those are satisfied then return. Something like -
> 
> ```
> void methodName() {
>        if (startSs.isHiveServerQuery) {
>            return;
>        }
>        if (!getSessionConf().getBoolVar(ConfVars.METASTORE_USE_THRIFT_SASL) {
>            return;
>        }
>        ...... (so on)
> }
> ```

Looks good, I will do it this way.

File: itests/hive-minikdc/src/test/java/org/apache/hadoop/hive/ql/session/TestAddMetastoreDelegationTokenToUGI.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (SourabhBadhya) [reply itests/hive-minikdc/src/test/java/org/apache/hadoop/hive/ql/session/TestAddMetastoreDelegationTokenToUGI.java:43]: Yes, adding tests in `itest/hive-minikdc/src/test/java/org/apache/hive/minikdc/TestRemoteHiveMetaStoreKerberos.java` is better.

File: ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java [hunk 8]
```diff
@@ -765,6 +768,32 @@ private static void start(SessionState startSs, boolean isAsync, LogHelper conso
...
+  private void setHiveMetastoreDelegationTokenToUGI() {
+    if (this.isHiveServerQuery) {
+      return;
+    }
+    if (!getSessionConf().getBoolVar(ConfVars.METASTORE_USE_THRIFT_SASL)) {
+      return;
+    }
+    try {
+      UserGroupInformation UGI = Utils.getUGI();
+      String hmsDelegationTokenStr = Hive.get().getDelegationToken(UGI.getShortUserName(), UGI.getShortUserName());
```
Comment (dengzhhu653) [inline ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java:783]: I think `Hive.get().getDelegationToken` will create a connection to metastore, and the session can reuse this Hive client later on. It looks the token got in this step takes no effect.

File: ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java [hunk 9]
```diff
@@ -765,6 +768,32 @@ private static void start(SessionState startSs, boolean isAsync, LogHelper conso
...
+      if (hmsDelegationTokenStr != null) {
+        getSessionConf().setVar(ConfVars.METASTORE_TOKEN_SIGNATURE, CLITOKEN);
+        try {
+          SecurityUtils.setTokenStr(UGI, hmsDelegationTokenStr, CLITOKEN);
```
Comment (dengzhhu653) [inline ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java:787]: In HiveSessionImplwithUGI, we set the HiveServer2ImpersonationToken token before the SessionState starts, will the HiveClientImpersonationToken override the HiveServer2ImpersonationToken?

File: ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java (hunk 8 above)
Comment (hzxiongyinke) [reply ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java:783]: > I think `Hive.get().getDelegationToken` will create a connection to metastore, and the session can reuse this Hive client later on. It looks the token got in this step takes no effect.

For iceberg issue https://github.com/apache/iceberg/issues/3127,insert into iceberg table with hive client ,iceberg table commit action occurs during the ApplicationMaster commit job phase without kerberos info,So we can only rely on HiveMetastore DelegationToken to solve this problem.

"will create a connection to metastore" means new HiveMetastoreClient , However, the hive client does not enable the "HADOOP_PROXY_USER", which means that no DelegationToken will be stuffed into the UGI.

File: ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java (hunk 9 above)
Comment (hzxiongyinke) [reply ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java:787]: > In HiveSessionImplwithUGI, we set the HiveServer2ImpersonationToken token before the SessionState starts, will the HiveClientImpersonationToken override the HiveServer2ImpersonationToken?

If hiveserverquery,will return directly

```
private void setHiveMetastoreDelegationTokenToUGI() {
    if (this.isHiveServerQuery) {
      return;
    }
...
}
```

File: ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java (hunk 8 above)
Comment (dengzhhu653) [reply ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java:783]: If this new method is called from the Kerberos yarn client(which is kerberos based), and want to pass the HMS token to yarn job, some thoughts:
1. what happen if the token being expired, should we get a new one? 
2. will the token in session config be propagated to the yarn job?

and I think we also need to check the client is kerberos based before fetching the dt: `UserGroupInformation.isSecurityEnabled()`
As this is Iceberg specific, an alternative way is to move this part to `configureJobConf`:
https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/metadata/HiveStorageHandler.java#L198

File: ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java (hunk 9 above)
Comment (dengzhhu653) [reply ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java:787]: Got it, thank you!

File: ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java (hunk 8 above)
Comment (hzxiongyinke) [reply ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java:783]: > If this new method is called from the Kerberos yarn client(which is kerberos based), and want to pass the HMS token to yarn job, some thoughts:
> 
> 1. what happen if the token being expired, should we get a new one?
> 2. will the token in session config be propagated to the yarn job?
> 
> and I think we also need to check the client is kerberos based before fetching the dt: `UserGroupInformation.isSecurityEnabled()` As this is Iceberg specific, an alternative way is to move this part to `configureJobConf`: https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/metadata/HiveStorageHandler.java#L198

1.if the token being expired , we can  do renew token such as  https://github.com/apache/hive/blob/698194183a2cb4ee6bc528ee71462cbae2c2c7ad/service/src/java/org/apache/hive/service/cli/session/HiveSessionImplwithUGI.java#L142-L161 .
i am not sure should we do this.
2.if token has set in UGI,new HMS client will check ConfVars.TOKEN_SIGNATURE for tokenSig.which ConfVars.TOKEN_SIGNATURE  has set in SessionState.so that we can connect to HMS (which is kerberos based) with HMS token

Comment (hzxiongyinke) [reply ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java:783]: > `hive.metastore.sasl.enabled = true` means HiveMetastore thrift interface will be secured with SASL. Clients must authenticate with Kerberos. so `UserGroupInformation.isSecurityEnabled()` is not necessary

File: ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java [hunk 10]
```diff
@@ -2302,4 +2331,6 @@ public Map<String, String> getLocalHdfsLocationMap(SessionState.ResourceType typ
     }
     return result;
   }
+
```
[the remark this answers is not in the document]
Comment (hzxiongyinke) [reply ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java]: fix

File: ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java (hunk 8 above)
Comment (dengzhhu653) [reply ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java:783]: Hi @hzxiongyinke,
> 1.if the token being expired , we can do renew token such as

I'm not sure how to trigger the token renew in HiveSessionImplWithUGI, as I understand from the context, this happens on Hive client, not HiveServer2.

> if token has set in UGI,new HMS client will check ConfVars.TOKEN_SIGNATURE for tokenSig.which ConfVars.TOKEN_SIGNATURE has set in SessionState

That's right, what I concern is that the App master doesn't sit inside the Hive client, they are two different processes, the App master may donot know the token in Hive client.

> Clients must authenticate with Kerberos

Right, this is a property for connecting to HMS, it doesn't mean the client itself is Kerberos based, I would prefer add the check here in my opinion.
